### ISSUE TITLE
Investigate different error recovery strategies

### ISSUE THREAD

**Issue by DmitrySoshnikov** [DmitrySoshnikov/syntax#28]
Currently the parser will fail on the first found error. For being more practical, we should consider building some [error recovery mechanisms](https://gist.github.com/DmitrySoshnikov/feee52cbfb03b7b691102ca240aa236a), and restore parsing after finding an error. We can start from the simplest _panic mode_ recovery, and allow catching several errors in one pass:

```
1 +& 2; 5 #* 3; 5 + 4;
```

Should issue two errors (currently issues only first one):

```
Unexpected token "&" at 1:3
Unexpected token "#" at 1:10
```

**Comment by hg42** [comment]
according to discussion in regexp-tree:

There are different kinds of parsing errors:
* fatal for parsing (like structural errors, often missing brackets etc.)
  - it is impossible to continue.
* fatal for parsing but with a good guess (like some kinds of missing semicolon or missing brackets)
  - this can be handled
  - sometimes by catch-all clauses
  - or closing error states at certain tokens like semicolons.
* errors with parsing values that are checked but don't influence parsing (like [z-a] in regexp)
  - this should not finish parsing at all (or better this would be configurable)
  - in many cases you want to get the AST for further processing (analyzing, reporting, fixing, etc.)

**Comment by DmitrySoshnikov** [comment]
> fatal for parsing but with a good guess (like some kinds of missing semicolon or missing brackets)

Yeah, this is exactly where we'd be doing recovery. The idea is again -- not to fail on first error, but try to catch as many parse errors as possible.

> errors with parsing values that are checked but don't influence parsing

This is not related to Syntax actually, and is up to specific parser (like regexp-tree parser). Syntax provides low-level parser generator API, not making any commitment about semantics of a particular language/grammar.
